From cc50855ca3ea21a6f4fb92e452914f5b05c6339c Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Wed, 15 Feb 2023 14:04:54 -0800
Subject: [PATCH] ath12k: Fix memory leak

During rmmod, insmod and in SSR case memory leak is observed.
1. Freed buff->payload in ath12k_dbring_buf_cleanup().
In ath12k_dp_cc_cleanup(), return is there after variable declaration
that don't allow func to do cleanup and memory is not getting freed.
So removed return.
2. Allocate arsta->rx_stats and ar->tx_stats if arsta->rx_stats and ar->tx_stats
is NULL respectively. rx_stats/tx_stats for each arsta is allocated when adding
a station. arsta->rx_stats and arsta->tx_stats will be freed when a station is
removed. ath12k_mac_station_add is called multiple times, so we need to stop
allocating rx_stats multiple times for same arsta.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dbring.c |  1 +
 drivers/net/wireless/ath/ath12k/mac.c    | 13 ++++++++-----
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dbring.c b/drivers/net/wireless/ath/ath12k/dbring.c
index 4445e42..7567e64 100644
--- a/drivers/net/wireless/ath/ath12k/dbring.c
+++ b/drivers/net/wireless/ath/ath12k/dbring.c
@@ -349,6 +349,7 @@ void ath12k_dbring_buf_cleanup(struct ath12k *ar, struct ath12k_dbring *ring)
 		idr_remove(&ring->bufs_idr, buf_id);
 		dma_unmap_single(ar->ab->dev, buff->paddr,
 				 ring->buf_sz, DMA_FROM_DEVICE);
+		kfree(buff->payload);
 		kfree(buff);
 	}
 
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index bdede83..484db40 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -5332,10 +5332,12 @@ static int ath12k_mac_station_add(struct ath12k *ar,
 		goto exit;
 	}
 
-	arsta->rx_stats = kzalloc(sizeof(*arsta->rx_stats), GFP_KERNEL);
 	if (!arsta->rx_stats) {
-		ret = -ENOMEM;
-		goto dec_num_station;
+		arsta->rx_stats = kzalloc(sizeof(*arsta->rx_stats), GFP_KERNEL);
+		if (!arsta->rx_stats) {
+			ret = -ENOMEM;
+			goto dec_num_station;
+		}
 	}
 
 	peer_param.vdev_id = arvif->vdev_id;
@@ -5352,8 +5354,9 @@ static int ath12k_mac_station_add(struct ath12k *ar,
 	ath12k_dbg(ab, ATH12K_DBG_MAC, "Added peer: %pM for VDEV: %d\n",
 		   sta->addr, arvif->vdev_id);
 
-	if (ath12k_debugfs_is_extd_tx_stats_enabled(ar)) {
-		arsta->tx_stats = kzalloc(sizeof(*arsta->tx_stats), GFP_KERNEL);
+	if (ath12k_debugfs_is_extd_tx_stats_enabled(ar) && (!arsta->tx_stats)) {
+		arsta->tx_stats = kzalloc(sizeof(*arsta->tx_stats),
+					  GFP_KERNEL);
 		if (!arsta->tx_stats) {
 			ret = -ENOMEM;
 			goto free_peer;
-- 
2.17.1

